pallets/author-mapping/src/lib.rs

```diff
@@ -32,29 +32,167 @@ pub use pallet::*;
+		#[pallet::weight(0)]
+		pub fn add_association(origin: OriginFor<T>, author_id: T::AuthorId) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			ensure!(
+				parachain_staking::Pallet::<T>::is_candidate(&account_id),
+				Error::<T>::CannotSetAuthor
+			);
+
+			T::DepositCurrency::reserve(&account_id, DEPOSIT_AMOUNT.into())
+				.map_err(|_| Error::<T>::CannotAffordSecurityDeposit)?;
+
+			Mapping::<T>::insert(&author_id, &account_id);
+
+			<Pallet<T>>::deposit_event(Event::AuthorRegistered(author_id, account_id));
+
+			Ok(())
+		}
+
+		/// Change your AuthorId.
+		///
+		/// This is useful for normal key rotation or for when switching from one pysical collator
+		/// machine to another. No new security deposit is required.
+		#[pallet::weight(0)]
+		pub fn update_association(
+			origin: OriginFor<T>,
+			old_author_id: T::AuthorId,
+			new_author_id: T::AuthorId,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			let stored_account = Mapping::<T>::try_get(&old_author_id)
+				.map_err(|_| Error::<T>::AssociationNotFound)?;
 
-		//TODO a call to clear your associated key and get your deposit back
+			ensure!(account_id == stored_account, Error::<T>::NotYourAssociation);
 
-		//TODO a call to clear someone else's defunt key? Maybe get a reward for doing it.
+			ensure!(
+				parachain_staking::Pallet::<T>::is_candidate(&account_id),
+				Error::<T>::CannotSetAuthor
+			);
+
+			Mapping::<T>::insert(&new_author_id, &account_id);
+
+			<Pallet<T>>::deposit_event(Event::AuthorRotated(new_author_id, account_id));
+
+			Ok(())
+		}
+
+		/// Clear your AuthorId.
+		///
+		/// This is useful when you are no longer an author and would like to re-claim your security
+		/// deposit.
+		#[pallet::weight(0)]
+		pub fn clear_association(
+			origin: OriginFor<T>,
+			author_id: T::AuthorId,
+		) -> DispatchResultWithPostInfo {
+			let account_id = ensure_signed(origin)?;
+
+			let stored_account =
+				Mapping::<T>::try_get(&author_id).map_err(|_| Error::<T>::AssociationNotFound)?;
+
+			ensure!(account_id == stored_account, Error::<T>::NotYourAssociation);
+
+			Mapping::<T>::remove(&author_id);
+
+			T::DepositCurrency::unreserve(&account_id, DEPOSIT_AMOUNT.into());
+
+			<Pallet<T>>::deposit_event(Event::AuthorDeRegistered(author_id));
+
+			Ok(().into())
+		}
+
+		/// Narc on another account for having a useless association and colelct a bounty.
+		///
+		/// This incentivizes good citizenship in the form of cleaning up others' defunct associations.
+		/// When you clean up another account's association, you will receive X percent of their
+		/// security deposit.
+		///
+		/// No association can be cleaned up within the initial grace period which allows collators
+		/// some time to get their associations onchain before they become active, and to clean up
+		/// after they are no longer active.
+		///
+		/// This also _forces_ collators to rotate their keys regularly because failing to do so will
```

Comment:
This is a bit risky no?
A collator who forgets to rotate its keys might end-up without having its block included.
How long where you thinking in terms of key rotation period ?

(Maybe this could be only be done on non-collating author)

Reply:
> A collator who forgets to rotate its keys might end-up without having its block included.

Yes, that's the design I'm advocating here. The idea is that proper key hygene is part of the responsibility of a collator. Collators who don't live up to their responsability miss out on the opportunity to author blocks and reap the benefits of doing so. I suspect (game theoretic analysis needed) that this will lead to a healthier collator community long term. Collators who don't follow proper key rotation will get their expired mappings narced and will miss rewards until they get new keys on chain. If a collator does this often, they will lose nominators in favor of better-performing collator candidates. This is all good.

That being said, I'm fine not including this mechanism as well. I think the incentive to get your security deposit back is enough to prevent state bloat.